pkg/dequeuer/dequeuer.go

```diff
@@ -0,0 +1,209 @@
+)
+
+type MessageHandler interface {
+	Handle(*sqs.Message) error
+}
+
+type SQSDequeuerConfig struct {
+	Region           string
+	QueueURL         string
+	StopIfNoMessages bool
+}
+
+type SQSDequeuer struct {
+	aws                *awslib.Client
+	config             SQSDequeuerConfig
+	hasDeadLetterQueue bool
+	waitTimeSeconds    *int64
+	visibilityTimeout  *int64
+	notFoundSleepTime  time.Duration
+	renewalPeriod      time.Duration
+	log                *zap.SugaredLogger
+	done               chan struct{}
+}
+
+func NewSQSDequeuer(config SQSDequeuerConfig, awsClient *awslib.Client, logger *zap.SugaredLogger) (*SQSDequeuer, error) {
+	attr, err := GetQueueAttributes(awsClient, config.QueueURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SQSDequeuer{
+		aws:                awsClient,
+		config:             config,
+		hasDeadLetterQueue: attr.RedrivePolicy,
+		waitTimeSeconds:    aws.Int64(int64(_waitTime.Seconds())),
+		visibilityTimeout:  aws.Int64(int64(_visibilityTimeout.Seconds())),
+		notFoundSleepTime:  _notFoundSleepTime,
+		renewalPeriod:      _renewalPeriod,
+		log:                logger,
+		done:               make(chan struct{}),
+	}, nil
+}
+
+func (d *SQSDequeuer) ReceiveMessage() ([]*sqs.Message, error) {
+	output, err := d.aws.SQS().ReceiveMessage(&sqs.ReceiveMessageInput{
+		QueueUrl:              aws.String(d.config.QueueURL),
+		MaxNumberOfMessages:   aws.Int64(_maxNumberOfMessages),
+		MessageAttributeNames: aws.StringSlice(_messageAttributes),
+		VisibilityTimeout:     d.visibilityTimeout,
+		WaitTimeSeconds:       d.waitTimeSeconds,
+	})
+
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return output.Messages, nil
+}
+
+func (d *SQSDequeuer) Start(messageHandler MessageHandler) error {
+	noMessagesInPreviousIteration := false
+
+loop:
+	for {
+		select {
+		case <-d.done:
+			break loop
+		default:
+			messages, err := d.ReceiveMessage()
+			if err != nil {
+				return err
+			}
+
+			if len(messages) == 0 {
+				queueAttributes, err := GetQueueAttributes(d.aws, d.config.QueueURL)
+				if err != nil {
+					return err
+				}
+
+				if queueAttributes.TotalMessages() == 0 {
+					if noMessagesInPreviousIteration && d.config.StopIfNoMessages {
+						d.log.Info("no messages found in queue, exiting ...")
+						return nil
+					}
+					noMessagesInPreviousIteration = true
+				}
+				time.Sleep(d.notFoundSleepTime)
+				continue
+			}
+
+			noMessagesInPreviousIteration = false
+			message := messages[0]
+			receiptHandle := *message.ReceiptHandle
+			done := d.StartMessageRenewer(receiptHandle)
+			d.handleMessage(message, messageHandler, done)
+		}
+	}
+
+	return nil
+}
+
+func (d *SQSDequeuer) Shutdown() {
+	d.done <- struct{}{}
+}
+
+func (d *SQSDequeuer) handleMessage(message *sqs.Message, messageHandler MessageHandler, done chan struct{}) {
+	messageErr := messageHandler.Handle(message)
+	// if messageErr != nil {
+	// 	// TODO: add to dead letter queue
+	// }
```

Comment:
Is this incomplete? (Same for the other TODOs in this function). If not, then we should add comments explaining why they are there and mention that it doesn't break the dequeuer.